UefiTestingPkg/FunctionalSystemTests/SmmPagingProtectionsTest/App/SmmPagingProtectionsTestApp.c

```diff
@@ -261,6 +288,322 @@ InvalidRangesShouldBeReadProtected (
 {
   EFI_STATUS  Status;
   BOOLEAN   PostReset = FALSE;
+  SMM_PAGING_PROTECTIONS_TEST_COMM_BUFFER *CommBuffer;
+
+  // Check to see whether we're loading a context, potentially after a reboot.
+  if (Context != NULL) {
+    PostReset = *(BOOLEAN*)Context;
+  }
+
+  // If we're not post-reset, this should be the first time this test runs.
+  if (!PostReset) {
+    UT_ASSERT_NOT_NULL (mPiSmmCommonCommBufferAddress);
+
+    //
+    // Since we expect the "test" code to cause a fault which will
+    // reset the system. Let's save a state that suggests the system
+    // has already reset. This way, when we resume we will consider
+    // it a "pass". If we fall through we will consider it a "fail".
+    //
+    PostReset = TRUE;
+    SetUsbBootNext ();
+    SaveFrameworkState (Framework, &PostReset, sizeof (PostReset));
+
+    // Grab the CommBuffer and fill it in for this test
+    Status = SmmMemoryProtectionsGetCommBuffer (&CommBuffer);
+    UT_ASSERT_NOT_EFI_ERROR (Status);
+
+    CommBuffer->Function = SMM_PAGING_PROTECTIONS_TEST_INVALID_RANGE;
+
+    // This should cause the system to reboot.
+    Status = SmmMemoryProtectionsDxeToSmmCommunicate ();
+
+    // If we're still here, things have gone wrong.
+    UT_LOG_ERROR ("System was expected to reboot, but didn't.");
+    PostReset = FALSE;
+    SaveFrameworkState (Framework, &PostReset, sizeof (PostReset));
+  }
+
+  UT_ASSERT_TRUE (PostReset);
+
+  return UNIT_TEST_PASSED;
+}
+
+UNIT_TEST_STATUS
+EFIAPI
+UnauthorizedIoShouldBeReadProtected (
+  IN UNIT_TEST_FRAMEWORK_HANDLE  Framework,
+  IN UNIT_TEST_CONTEXT           Context
+  )
+{
+  EFI_STATUS  Status;
+  BOOLEAN   PostReset = FALSE;
+  SMM_PAGING_PROTECTIONS_TEST_COMM_BUFFER *CommBuffer;
+
+  // Check to see whether we're loading a context, potentially after a reboot.
+  if (Context != NULL) {
+    PostReset = *(BOOLEAN*)Context;
+  }
+
+  // If we're not post-reset, this should be the first time this test runs.
+  if (!PostReset) {
+    UT_ASSERT_NOT_NULL (mPiSmmCommonCommBufferAddress);
+
+    //
+    // Since we expect the "test" code to cause a fault which will
+    // reset the system. Let's save a state that suggests the system
+    // has already reset. This way, when we resume we will consider
+    // it a "pass". If we fall through we will consider it a "fail".
+    //
+    PostReset = TRUE;
+    SetUsbBootNext ();
+    SaveFrameworkState (Framework, &PostReset, sizeof (PostReset));
+
+    // Grab the CommBuffer and fill it in for this test
+    Status = SmmMemoryProtectionsGetCommBuffer (&CommBuffer);
+    UT_ASSERT_NOT_EFI_ERROR (Status);
+
+    CommBuffer->Function = SMM_PROTECTIONS_READ_UNAUTHORIZED_IO;
```

Comment:
It seems like these "reset == success" functions could be rewritten as a single function that takes in the Function ID in the context. But maybe that would be more complicated later.

Reply:
Agreed. there's a lot of replicated "talk to SMM" boilerplate. But I didn't intend to rewrite the whole thing at the outset - just wanted to make myself satisfied that the new features worked as described. There is certainly room to rethink and refactor this whole thing in a vNext version.